File: Examples/Objective-C/FileShare/FileShare/Peer.m
//
//  Peer.m
//  FileShare
//
//  Created by Calvin on 6/9/17.
//  Copyright © 2017 Bridgefy Inc. All rights reserved.
//

#import "Peer.h"

@implementation Peer

- (id)initWithUUID:(NSString *)uuid {
    if (self = [super init]) {
        self.uuid = uuid;
        self.deviceType = DeviceTypeUndefined;
    }
    
    return self;
}

- (NSString *)formattedName {
    NSString *idFragment = [self.uuid substringToIndex:5];
    
    if (self.name)
        return [NSString stringWithFormat:@"%@ (%@)", self.name, idFragment];
    
    return idFragment;
}

- (void)createFiles:(NSArray *)files {
    self.files = [[NSMutableArray alloc] init];
    
    for (NSDictionary *fileDictionary in files) {
        FileInfo *fileInfo = [[FileInfo alloc] initWithDictionary:fileDictionary];
        
        [self.files addObject:fileInfo];
    }
}

- (FileInfo *)fileWithUUID:(NSString *)fileId {
    for (FileInfo *file in self.files) {
        if ([file.uuid isEqualToString:fileId]) {
            return file;
        }
    }
    
    return nil;
}

- (NSString *)description {
    return [NSString stringWithFormat:@"ID: %@\nName: %@\nType: %lu\nFiles: %@", self.uuid, self.name, (unsigned long)self.deviceType, self.files];
}

@end

File: Examples/Swift/FileShare/FileShare/Peer.swift
//
//  Peer.swift
//  FileShare
//
//  Created by Calvin on 6/20/17.
//  Copyright © 2017 Bridgefy Inc. All rights reserved.
//

import UIKit

enum DeviceType : Int {
    case undefined = 0
    case android
    case ios
}

class Peer: NSObject {
    
    var uuid: String
    var name: String?
    var deviceType: DeviceType
    var files: [FileInfo]
    
    required public init(uuid: String) {
        self.uuid = uuid
        self.deviceType = .undefined
        self.files = []
    }
    
    func formattedName() -> String {
        
        let index = self.uuid.index(self.uuid.startIndex, offsetBy: 5)
        let idFragment = self.uuid.substring(to: index)
        
        guard self.name != nil else {
            return idFragment
        }
        
        return "\(String(describing: self.name)) (\(idFragment))"
    }
    
    func createFiles(files: [[String: Any]]) {
        self.files = Array()
        
        for fileDictionary in files {
            let fileInfo = FileInfo(dictionary: fileDictionary)
            
            self.files.append(fileInfo!)
        }
    }
    
    func file(with uuid: String) -> FileInfo? {
        for file in self.files {
            if file.uuid == uuid {
                return file
            }
        }
        
        return nil
    }
    
    override var description: String {
        return "ID: \(self.uuid)\nName: \(self.name ?? "Not available")\nType: \(self.deviceType)\nFiles: \(String(describing: self.files))"
    }
    
}
